File: lib/opentelemetry_heroku.ex
defmodule OpentelemetryHeroku do
  @moduledoc """
  Implements an `OpenTelemetry` resource detector for Heroku.

  Exposes Heroku environment variables as an `OpenTelemetry` resource, using the
  semantic conventions defined in the `OpenTelemetry`
  [spec](https://opentelemetry.io/docs/specs/otel/resource/semantic_conventions/cloud_provider/heroku/).

  ## Usage:

  Add this module to the config:

  ```elixir
  config :opentelemetry, resource_detectors: [OpentelemetryHeroku]
  ```
  """
  require OpenTelemetry.SemanticConventions.Resource
  alias OpenTelemetry.SemanticConventions.Resource

  @behaviour :otel_resource_detector
  def get_resource(_) do
    [
      {Resource.cloud_provider(), "heroku"},
      {Resource.service_name(), System.get_env("HEROKU_APP_NAME")},
      {Resource.service_instance_id(), System.get_env("HEROKU_DYNO_ID")},
      {Resource.service_version(), System.get_env("HEROKU_RELEASE_VERSION")},
      {:"heroku.app.id", System.get_env("HEROKU_APP_ID")},
      {:"heroku.release.creation_timestamp", System.get_env("HEROKU_RELEASE_CREATED_AT")},
      {:"heroku.release.commit", System.get_env("HEROKU_SLUG_COMMIT")}
    ]
    |> :otel_resource.create()
  end
end

File: src/opentelemetry_heroku.erl
-module(opentelemetry_heroku).

-behaviour(otel_resource_detector).
-export([get_resource/1]).

-include_lib("opentelemetry_semantic_conventions/include/resource.hrl").

get_resource(_) ->
    Attributes = [
        {?CLOUD_PROVIDER, <<"heroku">>},
        {?SERVICE_NAME, get_env("HEROKU_APP_NAME")},
        {?SERVICE_INSTANCE_ID, get_env("HEROKU_DYNO_ID")},
        {?SERVICE_VERSION, get_env("HEROKU_RELEASE_VERSION")},
        {'heroku.app.id', get_env("HEROKU_APP_ID")},
        {'heroku.release.creation_timestamp', get_env("HEROKU_RELEASE_CREATED_AT")},
        {'heroku.release.commit', get_env("HEROKU_SLUG_COMMIT")}
    ],

    % Don't return environment variables that don't exist
    CleanAttributes = lists:filter(fun({_Key, Value}) -> is_binary(Value) end, Attributes),

    otel_resource:create(CleanAttributes).

get_env(Var) ->
    case os:getenv(Var) of
        false -> false;
        Val -> unicode:characters_to_binary(Val)
    end.
